product_operating_unit/models/product_template.py

```diff
@@ -18,18 +15,25 @@ def _default_operating_unit_ids(self):
                      self.env.uid).id])]
 
     operating_unit_ids = fields.Many2many(
-        'operating.unit', 'product_operating_unit_rel',
-        string='Operating Unit',
-        default=_default_operating_unit_ids
-    )
+        'operating.unit',
+        'product_operating_unit_rel',
+        string='Operating Units',
+        default=_default_operating_unit_ids)
+
+    @api.multi
+    @api.constrains('operating_unit_ids', 'categ_id')
+    def _check_operating_unit(self):
+        for record in self:
+            if record.categ_id.operating_unit_ids and \
+                    not all(ou in record.operating_unit_ids.ids for
+                            ou in record.categ_id.operating_unit_ids.ids):
+                raise ValidationError(_(
+                    "The operating units of the product must include the "
+                    "ones from the category."))
 
-    @api.constrains('operating_unit_ids', 'company_id')
-    def _check_company_operating_unit(self):
+    @api.multi
+    @api.onchange('categ_id')
+    def onchange_operating_unit_ids(self):
         for record in self:
-            if record.company_id and record.operating_unit_ids:
-                for operating_unit in record.operating_unit_ids:
-                    if record.company_id != operating_unit.company_id:
-                        raise ValidationError(
-                            _('Configuration error. The Company in the Product'
-                              ' and in the Operating Unit must be the same.')
-                        )
+            record.operating_unit_ids = \
```

Comment:
When changing the category you are resetting the ous to those coming from there but you are removing the possible other ones added for the product that differ from those of the category. Could it be changed so that it leaves those which weren't in the old category taking care of not duplicating them if they are in the new category?

Reply:
Yes, I agree @BT-rmartin  
When I removing OU in a category from one product but those categories are defined in other products than what should I Do?

Case 1:
category_1 have OU1 And OU2
Product_1 have category_1, so Product 1 has OU1 and OU2

Now, I removed category_1 in OU2
In product_2 have category_1, so Product 2 has only OU1

That Time Product_1 has OU1 and OU2, it's Right?

Reply:
You example is another use case. 
When you created Product1 the category_1 had OU1 and OU2. 
This means you weren't adding any OU in addition, but you just took the ones the category had. 
Now you are suggesting a new use case when you remove OU2 from category.

In this case you should update all products accordingly, not only new ones, but also existing ones. 
The reason is that even if you added it manually on the product when it wasn't in the category, in the end, you need to have added it to the category (and consequently to the product) to be able to remove it from the category (and consequently to the product).
 
Imagine the next scenario

Category_1: OU1 and OU2
Create Product1: It will have OU1 and OU2 from Category_1. Now imagine we add manually OU3
Edit Category_1: Remove OU2. Consequently you automatically removes from product OU2, which will keep OU1 and OU3
Edit Product1: You add OU2 manually, then it has again OU1, OU2 and OU3
Edit Category_1: You add OU2. Consequently it tries to add OU2 to the product, but it realizes it is already there, so no action
Edit Category_2: You remove OU2. Consequently you automatically remove from product OU2, but this time, it was added manually by user. Anyway it is removed as it might have been also added by previous change in Category_1, so the product will have OU1 and OU3

Then I would say that's ok. We can add and remove according to category operating units taken care of not creating duplicates and respecting the OUs added by the user not existing at any time in the category.